cou/steps/hypervisor.py

```diff
@@ -84,25 +113,40 @@ def apps(self) -> list[OpenStackApplication]:
         return self._apps
 
     @property
-    def azs(self) -> list[HypervisorGroup]:
+    def azs(self) -> AZs:
         """Returns a list of AZs defined in individual applications.
 
-        Each AZ contains a sorted list hypervisors, where each hypervisor has sorted list of units.
-        The order of units depends on the order in which they needs to be upgraded.
+        Each AZ contains a dictionary of application name and all units in the AZ.
         eg.
         az1:
-        - hypervisor0:
+        - cinder:
           - cinder/0
-          - nova-compute/0
-        - hypervisor1:
           - cinder/1
+          - cinder/2
+        - nova-compute
+          - nova-compute/0
           - nova-compute/1
+          - nova-compute/2
+        ...
+        az2
+        - cinder
+          -cinder/3
+        ...
         ...
 
-        :return: List of availability zones.
-        :rtype: list[HypervisorGroup]
+        :return: dictionary with key as az name and value as HypervisorGroup
+        :rtype: dict[str, HypervisorGroup]
+        :raises ApplicationError: if there is unit without az defined
         """
-        raise NotImplementedError
+        azs = AZs()
+        for app in self.apps:
+            for unit in app.units.values():
+                # NOTE(rgildein): If there is no AZ, we will use empty string and all units will
+                #                 belong to a single group.
+                az = unit.machine.az or ""
+                azs[az].app_units[app.name].append(unit)
+
+        return azs
```

Comment:
```suggestion
        az_groups = UpgradeGroups()
        for app in self.apps:
            for unit in app.units.values():
                # NOTE(rgildein): If there is no AZ, we will use empty string and all units will
                #                 belong to a single group.
                az = unit.machine.az or ""
                az_groups[az].app_units[app.name].append(unit)

        return az_groups
```

With something like this I think it's more agnostic to what is the logic behind the grouping. If in the future it needs to group with something different from azs it would still use the `UpgradeGroups` class, create a new property and populate the keys with whatever is needed.